{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Phenotypic Co-occurrence\n",
    "One measure enabled by the HPOA disease to phenotype data is phenotypic co-occurrence.  Phenotypic co-occurrence is of interest because it allows us to hypothesize that there is a dependent relationship between two phenotypes.  Although it does not imply causality, it may indicate a causal relationship between two phenotypes and and a third (or more) latent variables, or a causual relationship between two phenotypes. For example, _progressive muscle weakness_ causes _falls_.  Or an alternative example, there is a biological process and/or environmental factor that is causes allergies and asthma to co-occurr.  The latter is of interest in Monarch because we can query biological pathways and processes (GO, Reactome, etc.) related to phenotypes by joining gene-disease-phenotype relationships.  In addition, this type of analysis could  be useful to weight gene to phenotype associations where phenotypes co-occur in a mendelian disease where we have identified a causual gene to disease association.  This analysis could also be used to hypothesize pleiotropic effects.\n",
    "\n",
    "For simplicitiy, we will treat all phenotypes and diseases as flat (or leaf nodes) in our disease to phenotype association data.  In practice we know this is not correct, and both phenotype groups and disease groups appear in the association data.\n",
    "\n",
    "#### Approach\n",
    "We will query the HPOA association data using the Monarch Neo4J database.  These values may differ from querying the raw dataset due to merging equivalent diseases in MONDO.\n",
    "\n",
    "This analysis assumes we are starting with a phenotype of interest.  Although a more comprehensive analyses, as in generating a co-occurrence matrix, may also be useful.\n",
    "\n",
    "\n",
    "#### About this notebook\n",
    "This notebook uses the _Prominent nasal bridge_ as an example phenotype.  This can be changed in the second cell to analyze different phenotypes.\n",
    "\n",
    "Dependencies:\n",
    "\n",
    "* pip install requests\n",
    "* pip install pandas\n",
    "* pip install neo4j-driver"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 290,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+--------------------------------------------------+\n",
      "| phenotype                        | disease_count |\n",
      "+--------------------------------------------------+\n",
      "| \"Short stature\"                  | 67            |\n",
      "| \"Global developmental delay\"     | 67            |\n",
      "| \"Microcephaly\"                   | 67            |\n",
      "| \"Intellectual disability\"        | 65            |\n",
      "| \"Seizures\"                       | 50            |\n",
      "| \"Micrognathia\"                   | 48            |\n",
      "| \"Downslanted palpebral fissures\" | 46            |\n",
      "| \"High palate\"                    | 45            |\n",
      "| \"Cryptorchidism\"                 | 44            |\n",
      "| \"Low-set ears\"                   | 43            |\n",
      "+--------------------------------------------------+\n",
      "10 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "\n",
    "# Query the Monarch database for co-occurrence of prominent nasal bridge and other phenotypes\n",
    "# This can performed using a count aggregate function in cypher\n",
    "\n",
    "phenotype = \"HP:0000426\" # Prominent nasal bridge\n",
    "\n",
    "SCIGRAPH = \"https://scigraph-data-dev.monarchinitiative.org/scigraph/\"\n",
    "scigraph_exec = SCIGRAPH + \"cypher/execute\"\n",
    "scigraph_resolve = SCIGRAPH + \"cypher/resolve\"\n",
    "\n",
    "# The WHERE p1 <> p2 because clique merging of diseases causes\n",
    "# duplicate edges, need to fix\n",
    "cypher_query = \"\"\"\n",
    "    MATCH (disease:disease)-[:RO:0002200]->(p1:Node{iri:'%s'}),\n",
    "          (disease)-[:RO:0002200]->(p2:Node)\n",
    "    WHERE p1 <> p2\n",
    "    RETURN p2.label as phenotype, COUNT(DISTINCT(disease)) as disease_count\n",
    "    ORDER BY disease_count DESC\n",
    "\"\"\" % phenotype\n",
    "\n",
    "params = {\n",
    "    'cypherQuery': cypher_query,\n",
    "    'limit': 10\n",
    "}\n",
    "\n",
    "scigraph_req = requests.get(scigraph_exec, params=params)\n",
    "print(scigraph_req.text) # Default format is ascii text table, but we can get back json"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Normalization\n",
    "We know that the distribution of phenotypes across diseases is not uniform.  In other words, some phenotypes are annnotated to diseases more often than others.  Therefore, we need to normalize this data.  This is not to be confused with the frequency that a patient presents with a phenotype or expressivity.\n",
    "\n",
    "Two common methods for normalizing co-occurrence data are Jaccard similarity and Cosine similarity.\n",
    "\n",
    "We define these two approaches as:\n",
    "\n",
    "$$Jaccard(P1,P2) = \\frac{\\mid \\ P1 \\cap P2 \\ \\mid}{\\mid\\  P1 \\mid + \\mid P2 \\mid - \\mid P1 \\cap P2 \\ \\mid }$$\n",
    "\n",
    "Where \n",
    "$$ \\mid \\  P \\mid = \\text{Number of diseases annotated to phenotype P} $$\n",
    "\n",
    "For cosine similarity we will use the Ochiai coefficient, defined as:\n",
    "\n",
    "$$\\text{Ochiai coefficient(P1,P2)} = \\frac{\\mid P1 \\cap P2 \\mid}{\\sqrt{\\mid \\ P1 \\mid \\times \\mid P2 \\ \\mid}}$$\n",
    "\n",
    "#### Approach\n",
    "This can achieved in pure cypher, but we may consider creating a function on the server.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 291,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------------------------------------------------------------------------------------+\n",
      "| phenotype                        | p1_count | p2_count | intersection | normal_coef         |\n",
      "+---------------------------------------------------------------------------------------------+\n",
      "| \"Long face\"                      | 129      | 117      | 27           | 0.1232876712328767  |\n",
      "| \"Highly arched eyebrow\"          | 129      | 102      | 22           | 0.10526315789473684 |\n",
      "| \"Downslanted palpebral fissures\" | 129      | 372      | 46           | 0.1010989010989011  |\n",
      "| \"High palate\"                    | 129      | 429      | 45           | 0.08771929824561403 |\n",
      "| \"Thin vermilion border\"          | 129      | 110      | 19           | 0.08636363636363636 |\n",
      "| \"Posteriorly rotated ears\"       | 129      | 184      | 24           | 0.08304498269896193 |\n",
      "| \"Short philtrum\"                 | 129      | 151      | 21           | 0.08108108108108109 |\n",
      "| \"Macrotia\"                       | 129      | 191      | 24           | 0.08108108108108109 |\n",
      "| \"Narrow mouth\"                   | 129      | 210      | 25           | 0.07961783439490445 |\n",
      "| \"Sloping forehead\"               | 129      | 89       | 16           | 0.07920792079207921 |\n",
      "+---------------------------------------------------------------------------------------------+\n",
      "10 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Normalize with jaccard similarity\n",
    "cypher_query = \"\"\"\n",
    "    MATCH (disease:disease)-[:RO:0002200]->(p1:Node{iri:'%s'}),\n",
    "          (disease)-[:RO:0002200]->(p2:Phenotype)\n",
    "    WHERE p2 <> p1\n",
    "    WITH p2, COUNT(DISTINCT(disease)) as co_count\n",
    "    MATCH (disease:disease)-[:RO:0002200]->(p1:Node{iri:'%s'})\n",
    "    WITH COUNT(DISTINCT(disease)) as p1_count, p2, co_count\n",
    "    MATCH (disease:disease)-[:RO:0002200]->(p2)\n",
    "    WITH COUNT(DISTINCT(disease)) as p2_count, p1_count, p2, co_count\n",
    "    RETURN p2.label as phenotype, p1_count, p2_count, co_count as intersection,\n",
    "           toFloat(co_count)/((p1_count + p2_count)-co_count) as normal_coef\n",
    "    ORDER BY normal_coef DESC\n",
    "    \"\"\" % (phenotype, phenotype)\n",
    "\n",
    "params = {\n",
    "    'cypherQuery': cypher_query,\n",
    "    'limit': 10\n",
    "}\n",
    "\n",
    "scigraph_req = requests.get(scigraph_exec, params=params)\n",
    "print(scigraph_req.text) # Default format is ascii text table, but we can get back json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 292,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------------------------------------------------------------------------------------+\n",
      "| phenotype                        | p1_count | p2_count | intersection | normal_coef         |\n",
      "+---------------------------------------------------------------------------------------------+\n",
      "| \"Long face\"                      | 129      | 117      | 27           | 0.21977383072747697 |\n",
      "| \"Downslanted palpebral fissures\" | 129      | 372      | 46           | 0.20998656367149773 |\n",
      "| \"Microcephaly\"                   | 129      | 941      | 67           | 0.19230259103012684 |\n",
      "| \"Highly arched eyebrow\"          | 129      | 102      | 22           | 0.19179078635417854 |\n",
      "| \"High palate\"                    | 129      | 429      | 45           | 0.19128856646721754 |\n",
      "| \"Low-set ears\"                   | 129      | 488      | 43           | 0.1713814012647042  |\n",
      "| \"Short stature\"                  | 129      | 1187     | 67           | 0.17122003759530918 |\n",
      "| \"Ptosis\"                         | 129      | 499      | 42           | 0.16554046181613402 |\n",
      "| \"Micrognathia\"                   | 129      | 657      | 48           | 0.16487837460195826 |\n",
      "| \"Cryptorchidism\"                 | 129      | 584      | 44           | 0.16030659803408318 |\n",
      "+---------------------------------------------------------------------------------------------+\n",
      "10 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Normalize with ochiai coefficient\n",
    "cypher_query = \"\"\"\n",
    "    MATCH (disease:disease)-[:RO:0002200]->(p1:Node{iri:'%s'}),\n",
    "          (disease)-[:RO:0002200]->(p2:Phenotype)\n",
    "    WHERE p1 <> p2\n",
    "    WITH DISTINCT p1, p2, disease\n",
    "    WITH p2, COUNT(DISTINCT(disease)) as co_count\n",
    "    MATCH (disease:disease)-[:RO:0002200]->(p1:Node{iri:'%s'})\n",
    "    WITH COUNT(DISTINCT(disease)) as p1_count, p2, co_count\n",
    "    MATCH (disease:disease)-[:RO:0002200]->(p2)\n",
    "    WITH COUNT(DISTINCT(disease)) as p2_count, p1_count, p2, co_count\n",
    "    RETURN p2.label as phenotype, p1_count, p2_count, co_count as intersection,\n",
    "    toFloat(co_count)/sqrt(p1_count * p2_count) as normal_coef\n",
    "    ORDER BY normal_coef DESC\n",
    "    \"\"\" % (phenotype, phenotype)\n",
    "\n",
    "params = {\n",
    "    'cypherQuery': cypher_query,\n",
    "    'limit': 10\n",
    "}\n",
    "\n",
    "scigraph_req = requests.get(scigraph_exec, params=params)\n",
    "print(scigraph_req.text) # Default format is ascii text table, but we can get back json"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Results\n",
    "The original results showed several neurological abnormalities: seizures, global development delay, intellecutal disability.  In contrast, after normalization the majority of co-occurring phenotypes are related to morphilogical abnormalities of the head, with the exception of cryptorchidism.  It should be noted that cryptorchidism and microcephaly are phenotype groups, and considering disease to phenotype associations of their subclasses may downweight their co-occurrence.\n",
    "\n",
    "#### Normalization on phenotypic frequency\n",
    "For a single disease to phenotype association, the HPOAs provide a frequency field, defined as the frequency of patients that show a particular clinical feature. Examples are Obligate, Frequent, and Occasional.\n",
    "\n",
    "We can use this data to further weight/normalize phenotypic co-occurrence data.  For example, if two phenotypes occur \"very frequently\" in the same disease, this would be weighted higher than if one phenotype occurs very frequently, and one occurs occasionally.  We consider the intersection the minimum frequency between two phenotypes.  In addition, the total disease count will be adjusted to account for frequency.\n",
    "\n",
    "As a test we will set the following weights:\n",
    "\n",
    "| Frequency    | Definition     | Weight |\n",
    "|:-------------|:----------------|:--------|\n",
    "|Excluded      |present in 0%    | 0      |\n",
    "|Very rare     |present in 1-4%  |. 25    |\n",
    "|Occasional    |present in 5-29% |1.7    |\n",
    "|Frequent      |present in 30-79%|5.45    |\n",
    "|Very frequent |present in 80-99%|8.95    |\n",
    "|Obligate      |present in 100%|10    |\n",
    "|Not provided  ||4    |\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 293,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>disease</th>\n",
       "      <th>disease_curie</th>\n",
       "      <th>phenotype</th>\n",
       "      <th>phenotype_curie</th>\n",
       "      <th>phenotype_frequency</th>\n",
       "      <th>q_phenotype_frequency</th>\n",
       "      <th>qphenotype_curie</th>\n",
       "      <th>query_phenotype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>dentinogenesis imperfecta-short stature-hearin...</td>\n",
       "      <td>MONDO:0019102</td>\n",
       "      <td>Cone-shaped epiphysis</td>\n",
       "      <td>HP:0010579</td>\n",
       "      <td>8.95</td>\n",
       "      <td>8.95</td>\n",
       "      <td>HP:0000426</td>\n",
       "      <td>Prominent nasal bridge</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>dentinogenesis imperfecta-short stature-hearin...</td>\n",
       "      <td>MONDO:0019102</td>\n",
       "      <td>Intellectual disability, mild</td>\n",
       "      <td>HP:0001256</td>\n",
       "      <td>8.95</td>\n",
       "      <td>8.95</td>\n",
       "      <td>HP:0000426</td>\n",
       "      <td>Prominent nasal bridge</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>dentinogenesis imperfecta-short stature-hearin...</td>\n",
       "      <td>MONDO:0019102</td>\n",
       "      <td>Abnormal facial shape</td>\n",
       "      <td>HP:0001999</td>\n",
       "      <td>8.95</td>\n",
       "      <td>8.95</td>\n",
       "      <td>HP:0000426</td>\n",
       "      <td>Prominent nasal bridge</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>dentinogenesis imperfecta-short stature-hearin...</td>\n",
       "      <td>MONDO:0019102</td>\n",
       "      <td>Short philtrum</td>\n",
       "      <td>HP:0000322</td>\n",
       "      <td>8.95</td>\n",
       "      <td>8.95</td>\n",
       "      <td>HP:0000426</td>\n",
       "      <td>Prominent nasal bridge</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>dentinogenesis imperfecta-short stature-hearin...</td>\n",
       "      <td>MONDO:0019102</td>\n",
       "      <td>Sensorineural hearing impairment</td>\n",
       "      <td>HP:0000407</td>\n",
       "      <td>8.95</td>\n",
       "      <td>8.95</td>\n",
       "      <td>HP:0000426</td>\n",
       "      <td>Prominent nasal bridge</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             disease  disease_curie  \\\n",
       "0  dentinogenesis imperfecta-short stature-hearin...  MONDO:0019102   \n",
       "1  dentinogenesis imperfecta-short stature-hearin...  MONDO:0019102   \n",
       "2  dentinogenesis imperfecta-short stature-hearin...  MONDO:0019102   \n",
       "3  dentinogenesis imperfecta-short stature-hearin...  MONDO:0019102   \n",
       "4  dentinogenesis imperfecta-short stature-hearin...  MONDO:0019102   \n",
       "\n",
       "                          phenotype phenotype_curie  phenotype_frequency  \\\n",
       "0             Cone-shaped epiphysis      HP:0010579                 8.95   \n",
       "1     Intellectual disability, mild      HP:0001256                 8.95   \n",
       "2             Abnormal facial shape      HP:0001999                 8.95   \n",
       "3                    Short philtrum      HP:0000322                 8.95   \n",
       "4  Sensorineural hearing impairment      HP:0000407                 8.95   \n",
       "\n",
       "   q_phenotype_frequency qphenotype_curie         query_phenotype  \n",
       "0                   8.95       HP:0000426  Prominent nasal bridge  \n",
       "1                   8.95       HP:0000426  Prominent nasal bridge  \n",
       "2                   8.95       HP:0000426  Prominent nasal bridge  \n",
       "3                   8.95       HP:0000426  Prominent nasal bridge  \n",
       "4                   8.95       HP:0000426  Prominent nasal bridge  "
      ]
     },
     "execution_count": 293,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from neo4j.v1 import GraphDatabase\n",
    "\n",
    "# Frequency weight map\n",
    "freq_weights = {\n",
    "    'HP:0040285': 0,\n",
    "    'HP:0040284': .25,\n",
    "    'HP:0040283': 1.7,\n",
    "    'HP:0040282': 5.45,\n",
    "    'HP:0040281': 8.95,\n",
    "    'HP:0040280': 10,\n",
    "    'unknown':    4 # Not sure how to boost this\n",
    "}\n",
    "\n",
    "# Result table\n",
    "result_table = pd.DataFrame()\n",
    "\n",
    "# Note this query would be a lot shorter if frequencies were edge properties\n",
    "cypher_query = \"\"\"\n",
    "      MATCH (disease:disease)-[:RO:0002200]->(p1:Node{iri:'%s'}),\n",
    "            (disease)-[:RO:0002200]->(p2:Phenotype)\n",
    "      WHERE p1 <> p2\n",
    "      WITH DISTINCT p1, p2, disease\n",
    "      OPTIONAL MATCH (assoc:association)-[:OBAN:association_has_subject]->(disease),\n",
    "            (assoc)-[:OBAN:association_has_object]->(p1),\n",
    "            (assoc)-[:OBAN:association_has_predicate]->(has_pheno:Node{iri:'http://purl.obolibrary.org/obo/RO_0002200'}),\n",
    "            (assoc)-[::frequencyOfPhenotype]->(freq_p1)\n",
    "      WITH p1, p2, disease, freq_p1\n",
    "      OPTIONAL MATCH (assoc:association)-[:OBAN:association_has_subject]->(disease),\n",
    "            (assoc)-[:OBAN:association_has_object]->(p2),\n",
    "            (assoc)-[:OBAN:association_has_predicate]->(has_pheno:Node{iri:'http://purl.obolibrary.org/obo/RO_0002200'}),\n",
    "            (assoc)-[::frequencyOfPhenotype]->(freq_p2)\n",
    "      RETURN p1, p2, disease, freq_p1, freq_p2\n",
    "      \"\"\" % phenotype\n",
    "\n",
    "params = {\n",
    "    'cypherQuery': cypher_query\n",
    "}\n",
    "\n",
    "scigraph_req = requests.get(scigraph_resolve, params=params)\n",
    "resolved_query = scigraph_req.text # Resolve curies to IRIs\n",
    "\n",
    "scigraph_bolt = \"bolt://neo4j.monarchinitiative.org:443\"\n",
    "driver = GraphDatabase.driver(scigraph_bolt, auth=(\"neo4j\", \"password\"))\n",
    "\n",
    "def get_scigraph_results(query):\n",
    "    with driver.session() as session:\n",
    "        with session.begin_transaction() as tx:\n",
    "            for record in tx.run(query):\n",
    "                yield record\n",
    "\n",
    "for result in get_scigraph_results(resolved_query):\n",
    "    row = {}\n",
    "    row['query_phenotype'] = result['p1']['label']\n",
    "    row['phenotype'] = result['p2']['label']\n",
    "    row['disease'] = result['disease']['label']\n",
    "    row['qphenotype_curie'] = result['p1']['iri'].replace(\"http://purl.obolibrary.org/obo/HP_\", \"HP:\")\n",
    "    row['phenotype_curie'] = result['p2']['iri'].replace(\"http://purl.obolibrary.org/obo/HP_\", \"HP:\")\n",
    "    row['disease_curie'] = result['disease']['iri'].replace(\"http://purl.obolibrary.org/obo/MONDO_\", \"MONDO:\")\n",
    "\n",
    "    if result['freq_p1'] is not None:\n",
    "        freq_p1_curie = \\\n",
    "            result['freq_p1']['iri'].replace(\"http://purl.obolibrary.org/obo/HP_\", \"HP:\")\n",
    "    else:\n",
    "        freq_p1_curie = 'unknown'\n",
    "    if result['freq_p2'] is not None:\n",
    "        freq_p2_curie = \\\n",
    "            result['freq_p2']['iri'].replace(\"http://purl.obolibrary.org/obo/HP_\", \"HP:\")\n",
    "    else:\n",
    "        freq_p2_curie = 'unknown'\n",
    "        \n",
    "    row['q_phenotype_frequency'] = freq_weights[freq_p1_curie]\n",
    "    row['phenotype_frequency'] = freq_weights[freq_p2_curie]\n",
    "    \n",
    "    result_table = result_table.append(row, ignore_index=True)\n",
    "    \n",
    "result_table.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 294,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>intersection</th>\n",
       "      <th>phenotype</th>\n",
       "      <th>phenotype_curie</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>308.85</td>\n",
       "      <td>Global developmental delay</td>\n",
       "      <td>HP:0001263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>285.90</td>\n",
       "      <td>Microcephaly</td>\n",
       "      <td>HP:0000252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>285.90</td>\n",
       "      <td>Short stature</td>\n",
       "      <td>HP:0004322</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>275.35</td>\n",
       "      <td>Intellectual disability</td>\n",
       "      <td>HP:0001249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>213.90</td>\n",
       "      <td>Micrognathia</td>\n",
       "      <td>HP:0000347</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>212.20</td>\n",
       "      <td>Downslanted palpebral fissures</td>\n",
       "      <td>HP:0000494</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>192.50</td>\n",
       "      <td>Cryptorchidism</td>\n",
       "      <td>HP:0000028</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>185.75</td>\n",
       "      <td>High palate</td>\n",
       "      <td>HP:0000218</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>180.30</td>\n",
       "      <td>Low-set ears</td>\n",
       "      <td>HP:0000369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>176.40</td>\n",
       "      <td>Ptosis</td>\n",
       "      <td>HP:0000508</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     intersection                       phenotype phenotype_curie\n",
       "10         308.85      Global developmental delay      HP:0001263\n",
       "86         285.90                    Microcephaly      HP:0000252\n",
       "9          285.90                   Short stature      HP:0004322\n",
       "38         275.35         Intellectual disability      HP:0001249\n",
       "19         213.90                    Micrognathia      HP:0000347\n",
       "101        212.20  Downslanted palpebral fissures      HP:0000494\n",
       "18         192.50                  Cryptorchidism      HP:0000028\n",
       "17         185.75                     High palate      HP:0000218\n",
       "56         180.30                    Low-set ears      HP:0000369\n",
       "44         176.40                          Ptosis      HP:0000508"
      ]
     },
     "execution_count": 294,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aggregate_table = pd.DataFrame()\n",
    "\n",
    "phenotypes = result_table['phenotype'].unique()\n",
    "\n",
    "for pheno in phenotypes:\n",
    "    group_by_pheno = result_table[result_table['phenotype'] == pheno]\n",
    "    intersection = group_by_pheno.loc[:, ['phenotype_frequency', 'q_phenotype_frequency']].min(axis=1).sum()\n",
    "    row = {\n",
    "        'phenotype': pheno,\n",
    "        'phenotype_curie': group_by_pheno.iloc[0]['phenotype_curie'],\n",
    "        'intersection': intersection\n",
    "    }\n",
    "    aggregate_table = aggregate_table.append(row, ignore_index=True)\n",
    "\n",
    "aggregate_table.sort_values(by=['intersection'], ascending=False).head(10)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The top ten look similar to our original top ten list. However, we still need to normalize this data.  For the next step we will leverage our solr cache.  Solr/Golr is useful because we can toggle between treating phenotype disease annotations as flat or querying grouping classes (when applicable, such as microcephaly)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 295,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>intersection</th>\n",
       "      <th>phenotype</th>\n",
       "      <th>phenotype_curie</th>\n",
       "      <th>p2_count</th>\n",
       "      <th>p1_count</th>\n",
       "      <th>jaccard_sim</th>\n",
       "      <th>ochiai_coeff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>335</th>\n",
       "      <td>104.45</td>\n",
       "      <td>Long face</td>\n",
       "      <td>HP:0000276</td>\n",
       "      <td>517.25</td>\n",
       "      <td>638.55</td>\n",
       "      <td>0.099348</td>\n",
       "      <td>0.181744</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>212.20</td>\n",
       "      <td>Downslanted palpebral fissures</td>\n",
       "      <td>HP:0000494</td>\n",
       "      <td>1788.05</td>\n",
       "      <td>638.55</td>\n",
       "      <td>0.095827</td>\n",
       "      <td>0.198590</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>101.50</td>\n",
       "      <td>Short philtrum</td>\n",
       "      <td>HP:0000322</td>\n",
       "      <td>708.80</td>\n",
       "      <td>638.55</td>\n",
       "      <td>0.081470</td>\n",
       "      <td>0.150871</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>68.90</td>\n",
       "      <td>Low anterior hairline</td>\n",
       "      <td>HP:0000294</td>\n",
       "      <td>293.85</td>\n",
       "      <td>638.55</td>\n",
       "      <td>0.079792</td>\n",
       "      <td>0.159059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>185.75</td>\n",
       "      <td>High palate</td>\n",
       "      <td>HP:0000218</td>\n",
       "      <td>1889.00</td>\n",
       "      <td>638.55</td>\n",
       "      <td>0.079319</td>\n",
       "      <td>0.169128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>88.05</td>\n",
       "      <td>Thin vermilion border</td>\n",
       "      <td>HP:0000233</td>\n",
       "      <td>571.00</td>\n",
       "      <td>638.55</td>\n",
       "      <td>0.078511</td>\n",
       "      <td>0.145819</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>116.15</td>\n",
       "      <td>Macrotia</td>\n",
       "      <td>HP:0000400</td>\n",
       "      <td>970.85</td>\n",
       "      <td>638.55</td>\n",
       "      <td>0.077783</td>\n",
       "      <td>0.147518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>116.75</td>\n",
       "      <td>Narrow mouth</td>\n",
       "      <td>HP:0000160</td>\n",
       "      <td>1020.75</td>\n",
       "      <td>638.55</td>\n",
       "      <td>0.075686</td>\n",
       "      <td>0.144610</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>188</th>\n",
       "      <td>76.35</td>\n",
       "      <td>Highly arched eyebrow</td>\n",
       "      <td>HP:0002553</td>\n",
       "      <td>456.30</td>\n",
       "      <td>638.55</td>\n",
       "      <td>0.074963</td>\n",
       "      <td>0.141445</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>99.25</td>\n",
       "      <td>Posteriorly rotated ears</td>\n",
       "      <td>HP:0000358</td>\n",
       "      <td>795.50</td>\n",
       "      <td>638.55</td>\n",
       "      <td>0.074356</td>\n",
       "      <td>0.139256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>421</th>\n",
       "      <td>75.70</td>\n",
       "      <td>Underdeveloped nasal alae</td>\n",
       "      <td>HP:0000430</td>\n",
       "      <td>495.80</td>\n",
       "      <td>638.55</td>\n",
       "      <td>0.071506</td>\n",
       "      <td>0.134538</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>69.90</td>\n",
       "      <td>Sloping forehead</td>\n",
       "      <td>HP:0000340</td>\n",
       "      <td>436.25</td>\n",
       "      <td>638.55</td>\n",
       "      <td>0.069559</td>\n",
       "      <td>0.132438</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>164.10</td>\n",
       "      <td>Wide nasal bridge</td>\n",
       "      <td>HP:0000431</td>\n",
       "      <td>1885.90</td>\n",
       "      <td>638.55</td>\n",
       "      <td>0.069524</td>\n",
       "      <td>0.149538</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>95.45</td>\n",
       "      <td>Blepharophimosis</td>\n",
       "      <td>HP:0000581</td>\n",
       "      <td>837.10</td>\n",
       "      <td>638.55</td>\n",
       "      <td>0.069157</td>\n",
       "      <td>0.130554</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>83.45</td>\n",
       "      <td>Arachnodactyly</td>\n",
       "      <td>HP:0001166</td>\n",
       "      <td>653.30</td>\n",
       "      <td>638.55</td>\n",
       "      <td>0.069058</td>\n",
       "      <td>0.129203</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     intersection                       phenotype phenotype_curie  p2_count  \\\n",
       "335        104.45                       Long face      HP:0000276    517.25   \n",
       "101        212.20  Downslanted palpebral fissures      HP:0000494   1788.05   \n",
       "3          101.50                  Short philtrum      HP:0000322    708.80   \n",
       "160         68.90           Low anterior hairline      HP:0000294    293.85   \n",
       "17         185.75                     High palate      HP:0000218   1889.00   \n",
       "63          88.05           Thin vermilion border      HP:0000233    571.00   \n",
       "83         116.15                        Macrotia      HP:0000400    970.85   \n",
       "57         116.75                    Narrow mouth      HP:0000160   1020.75   \n",
       "188         76.35           Highly arched eyebrow      HP:0002553    456.30   \n",
       "147         99.25        Posteriorly rotated ears      HP:0000358    795.50   \n",
       "421         75.70       Underdeveloped nasal alae      HP:0000430    495.80   \n",
       "105         69.90                Sloping forehead      HP:0000340    436.25   \n",
       "49         164.10               Wide nasal bridge      HP:0000431   1885.90   \n",
       "48          95.45                Blepharophimosis      HP:0000581    837.10   \n",
       "13          83.45                  Arachnodactyly      HP:0001166    653.30   \n",
       "\n",
       "     p1_count  jaccard_sim  ochiai_coeff  \n",
       "335    638.55     0.099348      0.181744  \n",
       "101    638.55     0.095827      0.198590  \n",
       "3      638.55     0.081470      0.150871  \n",
       "160    638.55     0.079792      0.159059  \n",
       "17     638.55     0.079319      0.169128  \n",
       "63     638.55     0.078511      0.145819  \n",
       "83     638.55     0.077783      0.147518  \n",
       "57     638.55     0.075686      0.144610  \n",
       "188    638.55     0.074963      0.141445  \n",
       "147    638.55     0.074356      0.139256  \n",
       "421    638.55     0.071506      0.134538  \n",
       "105    638.55     0.069559      0.132438  \n",
       "49     638.55     0.069524      0.149538  \n",
       "48     638.55     0.069157      0.130554  \n",
       "13     638.55     0.069058      0.129203  "
      ]
     },
     "execution_count": 295,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import math\n",
    "\n",
    "# Pull down whole pivot table\n",
    "solr = 'https://solr-dev.monarchinitiative.org/solr/golr/select/'\n",
    "params = {\n",
    "  \"facet.pivot\": \"object,frequency\",\n",
    "  \"fq\": [\n",
    "    \"subject_category:disease\",\n",
    "    \"object_category:phenotype\"\n",
    "  ],\n",
    "  \"rows\": \"0\",\n",
    "  \"q\": \"*:*\",\n",
    "  \"facet.limit\": \"12000\", # Should get all of HPO\n",
    "  \"f.object_closure.facet.prefix\": \"HP\",\n",
    "  \"facet.method\": \"enum\",\n",
    "  \"facet.mincount\": \"1\",\n",
    "  \"facet\": \"true\",\n",
    "  \"wt\": \"json\",\n",
    "  \"facet.sort\": \"count\"\n",
    "};\n",
    "\n",
    "solr_req = requests.get(solr, params=params)\n",
    "pivot_table = solr_req.json()\n",
    "\n",
    "aggregate_table['p2_count'] = 0\n",
    "phenotype_ids = result_table['phenotype_curie'].unique()\n",
    "\n",
    "\n",
    "def calculate_weighted_frequency(facet):\n",
    "    count = (int(facet['count']) - sum([freq['count'] for freq in facet['pivot']])) * freq_weights['unknown']\n",
    "    for freq in facet['pivot']:\n",
    "        count += (freq['count'] * freq_weights[freq['value']])\n",
    "    return count\n",
    "\n",
    "for facet in pivot_table['facet_counts']['facet_pivot']['object,frequency']:\n",
    "    if facet['value'] == phenotype:\n",
    "        if 'pivot' in facet:\n",
    "            count = calculate_weighted_frequency(facet)\n",
    "        else:\n",
    "            count = int(facet['count']) * freq_weights['unknown']\n",
    "        aggregate_table['p1_count'] = count\n",
    "    elif facet['value'] in phenotype_ids:\n",
    "        if 'pivot' in facet:\n",
    "            count = calculate_weighted_frequency(facet)\n",
    "        else:\n",
    "            count = int(facet['count']) * freq_weights['unknown']\n",
    "        aggregate_table.loc[(aggregate_table['phenotype_curie'] == facet['value']), \"p2_count\"] = count\n",
    "\n",
    "def calculate_jaccard(intersection, count1, count2):\n",
    "    return intersection / ((count1 + count2) - intersection)\n",
    "\n",
    "def calcuate_ochiai(intersection, count1, count2):\n",
    "    return intersection / math.sqrt(count1 * count2)\n",
    "\n",
    "\n",
    "aggregate_table['jaccard_sim'] = aggregate_table.apply(\n",
    "        func=lambda row: calculate_jaccard(\n",
    "                              row['intersection'],\n",
    "                              row['p1_count'],\n",
    "                              row['p2_count']),\n",
    "        axis=1\n",
    ")\n",
    "    \n",
    "aggregate_table['ochiai_coeff'] = aggregate_table.apply(\n",
    "        func=lambda row: calcuate_ochiai(\n",
    "                              row['intersection'],\n",
    "                              row['p1_count'],\n",
    "                              row['p2_count']),\n",
    "        axis=1\n",
    ")\n",
    "\n",
    "aggregate_table.sort_values(by=['jaccard_sim'], ascending=False).head(15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 296,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>intersection</th>\n",
       "      <th>phenotype</th>\n",
       "      <th>phenotype_curie</th>\n",
       "      <th>p2_count</th>\n",
       "      <th>p1_count</th>\n",
       "      <th>jaccard_sim</th>\n",
       "      <th>ochiai_coeff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>212.20</td>\n",
       "      <td>Downslanted palpebral fissures</td>\n",
       "      <td>HP:0000494</td>\n",
       "      <td>1788.05</td>\n",
       "      <td>638.55</td>\n",
       "      <td>0.095827</td>\n",
       "      <td>0.198590</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>335</th>\n",
       "      <td>104.45</td>\n",
       "      <td>Long face</td>\n",
       "      <td>HP:0000276</td>\n",
       "      <td>517.25</td>\n",
       "      <td>638.55</td>\n",
       "      <td>0.099348</td>\n",
       "      <td>0.181744</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>285.90</td>\n",
       "      <td>Microcephaly</td>\n",
       "      <td>HP:0000252</td>\n",
       "      <td>4470.40</td>\n",
       "      <td>638.55</td>\n",
       "      <td>0.059278</td>\n",
       "      <td>0.169217</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>185.75</td>\n",
       "      <td>High palate</td>\n",
       "      <td>HP:0000218</td>\n",
       "      <td>1889.00</td>\n",
       "      <td>638.55</td>\n",
       "      <td>0.079319</td>\n",
       "      <td>0.169128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>68.90</td>\n",
       "      <td>Low anterior hairline</td>\n",
       "      <td>HP:0000294</td>\n",
       "      <td>293.85</td>\n",
       "      <td>638.55</td>\n",
       "      <td>0.079792</td>\n",
       "      <td>0.159059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>101.50</td>\n",
       "      <td>Short philtrum</td>\n",
       "      <td>HP:0000322</td>\n",
       "      <td>708.80</td>\n",
       "      <td>638.55</td>\n",
       "      <td>0.081470</td>\n",
       "      <td>0.150871</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113</th>\n",
       "      <td>168.50</td>\n",
       "      <td>Epicanthus</td>\n",
       "      <td>HP:0000286</td>\n",
       "      <td>1977.30</td>\n",
       "      <td>638.55</td>\n",
       "      <td>0.068850</td>\n",
       "      <td>0.149957</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>164.10</td>\n",
       "      <td>Wide nasal bridge</td>\n",
       "      <td>HP:0000431</td>\n",
       "      <td>1885.90</td>\n",
       "      <td>638.55</td>\n",
       "      <td>0.069524</td>\n",
       "      <td>0.149538</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>180.30</td>\n",
       "      <td>Low-set ears</td>\n",
       "      <td>HP:0000369</td>\n",
       "      <td>2288.95</td>\n",
       "      <td>638.55</td>\n",
       "      <td>0.065630</td>\n",
       "      <td>0.149135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>308.85</td>\n",
       "      <td>Global developmental delay</td>\n",
       "      <td>HP:0001263</td>\n",
       "      <td>6719.80</td>\n",
       "      <td>638.55</td>\n",
       "      <td>0.043812</td>\n",
       "      <td>0.149098</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>192.50</td>\n",
       "      <td>Cryptorchidism</td>\n",
       "      <td>HP:0000028</td>\n",
       "      <td>2629.50</td>\n",
       "      <td>638.55</td>\n",
       "      <td>0.062590</td>\n",
       "      <td>0.148558</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>176.40</td>\n",
       "      <td>Ptosis</td>\n",
       "      <td>HP:0000508</td>\n",
       "      <td>2218.15</td>\n",
       "      <td>638.55</td>\n",
       "      <td>0.065814</td>\n",
       "      <td>0.148220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>116.15</td>\n",
       "      <td>Macrotia</td>\n",
       "      <td>HP:0000400</td>\n",
       "      <td>970.85</td>\n",
       "      <td>638.55</td>\n",
       "      <td>0.077783</td>\n",
       "      <td>0.147518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>88.05</td>\n",
       "      <td>Thin vermilion border</td>\n",
       "      <td>HP:0000233</td>\n",
       "      <td>571.00</td>\n",
       "      <td>638.55</td>\n",
       "      <td>0.078511</td>\n",
       "      <td>0.145819</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>213.90</td>\n",
       "      <td>Micrognathia</td>\n",
       "      <td>HP:0000347</td>\n",
       "      <td>3373.95</td>\n",
       "      <td>638.55</td>\n",
       "      <td>0.056310</td>\n",
       "      <td>0.145728</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     intersection                       phenotype phenotype_curie  p2_count  \\\n",
       "101        212.20  Downslanted palpebral fissures      HP:0000494   1788.05   \n",
       "335        104.45                       Long face      HP:0000276    517.25   \n",
       "86         285.90                    Microcephaly      HP:0000252   4470.40   \n",
       "17         185.75                     High palate      HP:0000218   1889.00   \n",
       "160         68.90           Low anterior hairline      HP:0000294    293.85   \n",
       "3          101.50                  Short philtrum      HP:0000322    708.80   \n",
       "113        168.50                      Epicanthus      HP:0000286   1977.30   \n",
       "49         164.10               Wide nasal bridge      HP:0000431   1885.90   \n",
       "56         180.30                    Low-set ears      HP:0000369   2288.95   \n",
       "10         308.85      Global developmental delay      HP:0001263   6719.80   \n",
       "18         192.50                  Cryptorchidism      HP:0000028   2629.50   \n",
       "44         176.40                          Ptosis      HP:0000508   2218.15   \n",
       "83         116.15                        Macrotia      HP:0000400    970.85   \n",
       "63          88.05           Thin vermilion border      HP:0000233    571.00   \n",
       "19         213.90                    Micrognathia      HP:0000347   3373.95   \n",
       "\n",
       "     p1_count  jaccard_sim  ochiai_coeff  \n",
       "101    638.55     0.095827      0.198590  \n",
       "335    638.55     0.099348      0.181744  \n",
       "86     638.55     0.059278      0.169217  \n",
       "17     638.55     0.079319      0.169128  \n",
       "160    638.55     0.079792      0.159059  \n",
       "3      638.55     0.081470      0.150871  \n",
       "113    638.55     0.068850      0.149957  \n",
       "49     638.55     0.069524      0.149538  \n",
       "56     638.55     0.065630      0.149135  \n",
       "10     638.55     0.043812      0.149098  \n",
       "18     638.55     0.062590      0.148558  \n",
       "44     638.55     0.065814      0.148220  \n",
       "83     638.55     0.077783      0.147518  \n",
       "63     638.55     0.078511      0.145819  \n",
       "19     638.55     0.056310      0.145728  "
      ]
     },
     "execution_count": 296,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aggregate_table.sort_values(by=['ochiai_coeff'], ascending=False).head(15)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "For this example, adding frequency data does not seem to affect the results much with the exception that _low anterior hairline_ is upranked in the frequency aware lists.\n",
    "\n",
    "Follow up questions:\n",
    "- What cutoff is considered significant for grouping phenotypes (or should we generate a matrix and cluster?)\n",
    "- What genes are associated with this group of phenotypes?\n",
    "- What processes and pathways are associated with these genes?\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 297,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['OTUD6B',\n",
       " 'MED12',\n",
       " 'TMEM237',\n",
       " 'CEP152',\n",
       " 'FBN1',\n",
       " 'TBX1',\n",
       " 'CEP290',\n",
       " 'SNAP29',\n",
       " 'SIN3A',\n",
       " 'EBF3',\n",
       " 'MKS1',\n",
       " 'HERC1',\n",
       " 'RECQL4',\n",
       " 'NPHP1',\n",
       " 'SATB2',\n",
       " 'BDH1',\n",
       " 'KAT6B']"
      ]
     },
     "execution_count": 297,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get the top two hits\n",
    "top_phenotypes = aggregate_table.sort_values(\n",
    "    by=['jaccard_sim'],\n",
    "    ascending=False)['phenotype_curie'][0:2]\n",
    "\n",
    "# Add query phenotype\n",
    "pheno_group = top_phenotypes.tolist()\n",
    "pheno_group.append(phenotype)\n",
    "\n",
    "def get_solr_docs(solr, params):\n",
    "    resultCount = params['rows']\n",
    "    while params['start'] < resultCount:\n",
    "        solr_request = requests.get(solr, params=params)\n",
    "        response = solr_request.json()\n",
    "        resultCount = response['response']['numFound']\n",
    "        params['start'] += params['rows']\n",
    "        for doc in response['response']['docs']:\n",
    "            yield doc\n",
    "            \n",
    "# Get intersection of genes\n",
    "def get_gene_intersect(phenotypes, solr):\n",
    "    genes = dict()\n",
    "    is_first = True\n",
    "    for term in phenotypes:\n",
    "        params = {\n",
    "            'wt': 'json',\n",
    "            'rows': 500,\n",
    "            'start': 0,\n",
    "            'q': '*:*',\n",
    "            'fl': 'subject, subject_label',\n",
    "            'fq': ['subject_category:\"gene\"',\n",
    "                   'object_closure: \"{}\"'.format(term)\n",
    "                  ]\n",
    "        }\n",
    "        temp_dict = {}\n",
    "        for doc in get_solr_docs(solr, params):\n",
    "            temp_dict[doc['subject']] = doc['subject_label']\n",
    "        \n",
    "        if is_first:\n",
    "            genes = temp_dict\n",
    "        else:\n",
    "            gene_set = genes.keys() & temp_dict.keys()\n",
    "            genes = {k: temp_dict[k] for k in gene_set}\n",
    "        is_first = False\n",
    "    return genes\n",
    "\n",
    "genes = get_gene_intersect(pheno_group, solr)\n",
    "\n",
    "list(genes.values())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 330,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'AURKA Activation by TPX2': 4,\n",
       " 'Adaptive Immune System': 1,\n",
       " 'Anchoring of the basal body to the plasma membrane': 8,\n",
       " 'Antigen processing: Ubiquitination & Proteasome degradation': 2,\n",
       " 'Autophagy - animal': 1,\n",
       " 'Butanoate metabolism': 1,\n",
       " 'Cell Cycle': 2,\n",
       " 'Cell Cycle, Mitotic': 2,\n",
       " 'Centrosome maturation': 2,\n",
       " 'Chromatin modifying enzymes': 1,\n",
       " 'Chromatin organization': 1,\n",
       " 'Cilium Assembly': 4,\n",
       " 'Class I MHC mediated antigen processing & presentation': 1,\n",
       " 'DNA duplex unwinding': 1,\n",
       " 'DNA repair': 1,\n",
       " 'DNA replication': 2,\n",
       " 'DNA strand renaturation': 1,\n",
       " 'Degradation of the extracellular matrix': 2,\n",
       " 'Developmental Biology': 1,\n",
       " 'Elastic fibre formation': 2,\n",
       " 'Epigenetic regulation of gene expression': 1,\n",
       " 'Extracellular matrix organization': 1,\n",
       " 'Factors involved in megakaryocyte development and platelet production': 2,\n",
       " 'Fatty acid, triacylglycerol, and ketone body metabolism': 3,\n",
       " 'G2/M Transition': 2,\n",
       " 'G2/M transition of mitotic cell cycle': 2,\n",
       " 'Gene Expression': 2,\n",
       " 'Generic Transcription Pathway': 2,\n",
       " 'HATs acetylate histones': 2,\n",
       " \"Hedgehog 'off' state\": 2,\n",
       " 'Hemostasis': 1,\n",
       " \"Huntington's disease\": 1,\n",
       " 'Immune System': 3,\n",
       " 'Innate Immune System': 2,\n",
       " 'Integrin cell surface interactions': 2,\n",
       " 'Intra-Golgi and retrograde Golgi-to-ER traffic': 1,\n",
       " 'Intra-Golgi traffic': 2,\n",
       " 'Ketone body metabolism': 1,\n",
       " 'Loss of Nlp from mitotic centrosomes': 4,\n",
       " 'Loss of proteins required for interphase microtubule organization from the centrosome': 4,\n",
       " 'Membrane Trafficking': 1,\n",
       " 'Metabolic pathways': 1,\n",
       " 'Metabolism': 3,\n",
       " 'Metabolism of lipids and lipoproteins': 3,\n",
       " 'Mitotic G2-G2/M phases': 2,\n",
       " 'Molecules associated with elastic fibres': 2,\n",
       " 'Negative epigenetic regulation of rRNA expression': 1,\n",
       " 'Neutrophil degranulation': 4,\n",
       " 'NoRC negatively regulates rRNA expression': 2,\n",
       " 'Organelle biogenesis and maintenance': 4,\n",
       " 'PPARA activates gene expression': 2,\n",
       " 'Post-translational protein phosphorylation': 2,\n",
       " 'RUNX1 regulates genes involved in megakaryocyte differentiation and platelet function': 2,\n",
       " 'RUNX2 regulates osteoblast differentiation': 2,\n",
       " 'Recruitment of NuMA to mitotic centrosomes': 4,\n",
       " 'Recruitment of mitotic centrosome proteins and complexes': 4,\n",
       " 'Regulation of Insulin-like Growth Factor (IGF) transport and uptake by Insulin-like Growth Factor Binding Proteins (IGFBPs)': 2,\n",
       " 'Regulation of PLK1 Activity at G2/M Transition': 4,\n",
       " 'Regulation of lipid metabolism by Peroxisome proliferator-activated receptor alpha (PPARalpha)': 3,\n",
       " 'SNARE interactions in vesicular transport': 1,\n",
       " 'SUMOylation of chromatin organization proteins': 2,\n",
       " 'Schwann cell development': 1,\n",
       " 'Signal Transduction': 1,\n",
       " 'Signaling by Hedgehog': 1,\n",
       " 'Synthesis and degradation of ketone bodies': 1,\n",
       " 'Synthesis of Ketone Bodies': 2,\n",
       " 'Thyroid hormone signaling pathway': 2,\n",
       " 'Transcriptional misregulation in cancer': 1,\n",
       " 'Transcriptional regulation of white adipocyte differentiation': 2,\n",
       " 'Ubiquitin mediated proteolysis': 1,\n",
       " 'Utilization of Ketone Bodies': 2,\n",
       " 'Vesicle-mediated transport': 1,\n",
       " 'Wnt signaling pathway, planar cell polarity pathway': 1,\n",
       " 'actin cytoskeleton organization': 1,\n",
       " 'activation of innate immune response': 1,\n",
       " 'aging': 1,\n",
       " 'androgen receptor signaling pathway': 1,\n",
       " 'angiogenesis': 1,\n",
       " 'anterior/posterior pattern specification': 1,\n",
       " 'aorta morphogenesis': 1,\n",
       " 'artery morphogenesis': 1,\n",
       " 'autophagosome maturation': 1,\n",
       " 'autophagosome membrane docking': 1,\n",
       " 'axis elongation involved in somitogenesis': 1,\n",
       " 'base-excision repair': 1,\n",
       " 'blood vessel development': 1,\n",
       " 'blood vessel morphogenesis': 1,\n",
       " 'branching morphogenesis of an epithelial tube': 1,\n",
       " 'camera-type eye development': 1,\n",
       " 'canonical Wnt signaling pathway': 1,\n",
       " 'cartilage development': 1,\n",
       " 'cell adhesion mediated by integrin': 1,\n",
       " 'cell fate specification': 1,\n",
       " 'cell projection organization': 1,\n",
       " 'cell proliferation': 2,\n",
       " 'cell-cell adhesion': 1,\n",
       " 'cellular protein localization': 1,\n",
       " 'cellular protein metabolic process': 1,\n",
       " 'cellular response to dopamine': 1,\n",
       " 'cellular response to fibroblast growth factor stimulus': 1,\n",
       " 'cellular response to glucose stimulus': 1,\n",
       " 'cellular response to insulin-like growth factor stimulus': 1,\n",
       " 'cellular response to leukemia inhibitory factor': 1,\n",
       " 'cellular response to organic substance': 1,\n",
       " 'cellular response to retinoic acid': 1,\n",
       " 'cellular response to transforming growth factor beta stimulus': 1,\n",
       " 'centriole replication': 1,\n",
       " 'centrosome duplication': 1,\n",
       " 'cerebellar Purkinje cell differentiation': 1,\n",
       " 'chromatin remodeling': 1,\n",
       " 'ciliary basal body-plasma membrane docking': 4,\n",
       " 'cilium assembly': 4,\n",
       " 'cochlea morphogenesis': 1,\n",
       " 'commitment of neuronal cell to specific neuron type in forebrain': 1,\n",
       " 'common bile duct development': 1,\n",
       " 'coronary artery morphogenesis': 1,\n",
       " 'covalent chromatin modification': 1,\n",
       " 'de novo centriole assembly involved in multi-ciliated epithelial cell differentiation': 1,\n",
       " 'determination of left/right symmetry': 2,\n",
       " 'double-strand break repair via homologous recombination': 1,\n",
       " 'ear morphogenesis': 1,\n",
       " 'embryonic brain development': 2,\n",
       " 'embryonic cranial skeleton morphogenesis': 1,\n",
       " 'embryonic digit morphogenesis': 1,\n",
       " 'embryonic eye morphogenesis': 1,\n",
       " 'embryonic neurocranium morphogenesis': 1,\n",
       " 'embryonic pattern specification': 1,\n",
       " 'embryonic skeletal system development': 1,\n",
       " 'embryonic skeletal system morphogenesis': 1,\n",
       " 'embryonic viscerocranium morphogenesis': 1,\n",
       " 'enamel mineralization': 1,\n",
       " 'endoderm development': 1,\n",
       " 'epithelial cell differentiation': 1,\n",
       " 'epithelial structure maintenance': 1,\n",
       " 'excretion': 1,\n",
       " 'exocytosis': 1,\n",
       " 'extracellular matrix disassembly': 1,\n",
       " 'extracellular matrix organization': 1,\n",
       " 'eye photoreceptor cell development': 1,\n",
       " 'face morphogenesis': 1,\n",
       " 'head development': 1,\n",
       " 'heart development': 3,\n",
       " 'heart morphogenesis': 1,\n",
       " 'hematopoietic progenitor cell differentiation': 1,\n",
       " 'hindbrain development': 1,\n",
       " 'histone H3 acetylation': 1,\n",
       " 'histone acetylation': 1,\n",
       " 'histone deacetylation': 1,\n",
       " 'in utero embryonic development': 1,\n",
       " 'inner ear morphogenesis': 1,\n",
       " 'inner ear receptor cell stereocilium organization': 1,\n",
       " 'intracellular steroid hormone receptor signaling pathway': 1,\n",
       " 'ketone body biosynthetic process': 1,\n",
       " 'ketone body catabolic process': 1,\n",
       " 'lymph vessel development': 1,\n",
       " 'membrane fusion': 1,\n",
       " 'mesenchymal cell apoptotic process': 1,\n",
       " 'mesoderm development': 1,\n",
       " 'metanephros development': 1,\n",
       " 'middle ear morphogenesis': 1,\n",
       " 'motile cilium assembly': 1,\n",
       " 'multicellular organism development': 2,\n",
       " 'muscle cell fate commitment': 1,\n",
       " 'muscle organ development': 1,\n",
       " 'muscle organ morphogenesis': 1,\n",
       " 'muscle tissue morphogenesis': 1,\n",
       " 'negative regulation of Wnt signaling pathway': 1,\n",
       " 'negative regulation of apoptotic process': 1,\n",
       " 'negative regulation of autophagy': 1,\n",
       " 'negative regulation of cell differentiation': 1,\n",
       " 'negative regulation of circadian rhythm': 1,\n",
       " 'negative regulation of histone H3-K27 acetylation': 1,\n",
       " 'negative regulation of mesenchymal cell apoptotic process': 1,\n",
       " 'negative regulation of osteoclast development': 1,\n",
       " 'negative regulation of osteoclast differentiation': 1,\n",
       " 'negative regulation of protein localization to nucleus': 1,\n",
       " 'negative regulation of transcription by RNA polymerase II': 2,\n",
       " 'negative regulation of transcription regulatory region DNA binding': 1,\n",
       " 'negative regulation of transcription, DNA-templated': 2,\n",
       " 'neural crest cell migration': 1,\n",
       " 'neural tube closure': 2,\n",
       " 'neuromuscular process controlling balance': 1,\n",
       " 'neuron migration': 1,\n",
       " 'neuron projection development': 1,\n",
       " 'neutrophil degranulation': 2,\n",
       " 'non-motile cilium assembly': 1,\n",
       " 'nucleosome assembly': 1,\n",
       " 'odontogenesis of dentin-containing tooth': 1,\n",
       " 'oligodendrocyte development': 1,\n",
       " 'osteoblast development': 1,\n",
       " 'otic vesicle formation': 1,\n",
       " 'outer ear morphogenesis': 1,\n",
       " 'outflow tract morphogenesis': 1,\n",
       " 'outflow tract septum morphogenesis': 1,\n",
       " 'oxidation-reduction process': 1,\n",
       " 'palate development': 1,\n",
       " 'parathyroid gland development': 1,\n",
       " 'pattern specification process': 1,\n",
       " 'pharyngeal system development': 1,\n",
       " 'positive regulation of G2/M transition of mitotic cell cycle': 1,\n",
       " 'positive regulation of MAPK cascade': 1,\n",
       " 'positive regulation of bicellular tight junction assembly': 1,\n",
       " 'positive regulation of cell proliferation': 1,\n",
       " 'positive regulation of chromatin silencing': 1,\n",
       " 'positive regulation of defense response to virus by host': 1,\n",
       " 'positive regulation of epithelial cell proliferation': 1,\n",
       " 'positive regulation of intracellular protein transport': 1,\n",
       " 'positive regulation of mesenchymal cell proliferation': 1,\n",
       " 'positive regulation of protein phosphorylation': 1,\n",
       " 'positive regulation of tongue muscle cell differentiation': 1,\n",
       " 'positive regulation of transcription from RNA polymerase II promoter': 6,\n",
       " 'positive regulation of transcription, DNA-templated': 5,\n",
       " 'post-anal tail morphogenesis': 1,\n",
       " 'post-embryonic eye morphogenesis': 1,\n",
       " 'post-translational protein modification': 1,\n",
       " 'pronephros development': 1,\n",
       " 'protein deacetylation': 1,\n",
       " 'protein deubiquitination': 1,\n",
       " 'protein transport': 2,\n",
       " 'protein ubiquitination': 2,\n",
       " 'regulation of G2/M transition of mitotic cell cycle': 2,\n",
       " 'regulation of Wnt signaling pathway': 1,\n",
       " 'regulation of Wnt signaling pathway, planar cell polarity pathway': 1,\n",
       " 'regulation of canonical Wnt signaling pathway': 1,\n",
       " 'regulation of cellular response to growth factor stimulus': 1,\n",
       " 'regulation of establishment of protein localization': 1,\n",
       " 'regulation of hormone levels': 1,\n",
       " 'regulation of lipid metabolic process': 1,\n",
       " 'regulation of megakaryocyte differentiation': 1,\n",
       " 'regulation of molecular function': 1,\n",
       " 'regulation of organ morphogenesis': 1,\n",
       " 'regulation of receptor activity': 1,\n",
       " 'regulation of smoothened signaling pathway involved in dorsal/ventral neural tube patterning': 1,\n",
       " 'regulation of transcription from RNA polymerase II promoter': 2,\n",
       " 'regulation of transcription from RNA polymerase II promoter in response to oxidative stress': 1,\n",
       " 'regulation of transcription, DNA-templated': 1,\n",
       " 'response to methylglyoxal': 1,\n",
       " 'retina development in camera-type eye': 1,\n",
       " 'retinoic acid receptor signaling pathway': 1,\n",
       " 'rhythmic process': 1,\n",
       " 'semicircular canal morphogenesis': 1,\n",
       " 'sensory perception of sound': 1,\n",
       " 'sequestering of BMP in extracellular matrix': 1,\n",
       " 'sequestering of TGFbeta in extracellular matrix': 1,\n",
       " 'signal transduction': 1,\n",
       " 'skeletal system development': 1,\n",
       " 'social behavior': 1,\n",
       " 'soft palate development': 1,\n",
       " 'spermatid differentiation': 1,\n",
       " 'spinal cord development': 1,\n",
       " 'stem cell population maintenance': 1,\n",
       " 'synaptic vesicle fusion to presynaptic active zone membrane': 1,\n",
       " 'synaptic vesicle priming': 1,\n",
       " 'telomere maintenance': 1,\n",
       " 'telomeric D-loop disassembly': 1,\n",
       " 'thymus development': 1,\n",
       " 'thyroid gland development': 1,\n",
       " 'tongue morphogenesis': 1,\n",
       " 'transcription from RNA polymerase II promoter': 2,\n",
       " 'transcription initiation from RNA polymerase II promoter': 1,\n",
       " 'transcription, DNA-templated': 3,\n",
       " 'vagus nerve morphogenesis': 1,\n",
       " 'vesicle targeting': 1,\n",
       " 'visual behavior': 1}"
      ]
     },
     "execution_count": 330,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import itertools\n",
    "# Get all related pathways and GO processes\n",
    "\n",
    "def get_facet_table(solr, params):\n",
    "    \"\"\"\n",
    "    fetch and convert facet table to dictionary\n",
    "    \"\"\"\n",
    "    results = {}\n",
    "    solr_request = requests.get(solr, params=params)\n",
    "    response = solr_request.json()\n",
    "    for facet in response['facet_counts']['facet_fields'].values():\n",
    "        temp_dict = dict(itertools.zip_longest(*[iter(facet)] * 2, fillvalue=\"\"))\n",
    "        results = {**results, **temp_dict}\n",
    "    return results\n",
    "\n",
    "gene_or_filter = \"\"\n",
    "for gene in genes.keys():\n",
    "    gene_or_filter += 'OR subject_closure:\"{}\" '.format(gene)\n",
    "gene_or_filter = gene_or_filter[3:]\n",
    "\n",
    "# Go mol function params\n",
    "params = {\n",
    "    'wt': 'json',\n",
    "    'rows': 0,\n",
    "    'q': '*:*',\n",
    "    'facet': 'true',\n",
    "    'facet.sort': 'count',\n",
    "    'facet.method': 'enum',\n",
    "    'facet.mincount': '1',\n",
    "    'facet.limit': '3000',\n",
    "    'facet.field': 'object_label',\n",
    "    'fl': 'subject, subject_label',\n",
    "    'fq': [gene_or_filter,\n",
    "           'relation:\"RO:0002331\"',\n",
    "           \n",
    "          ]\n",
    "}\n",
    "\n",
    "go_functions = get_facet_table(solr, params)\n",
    "go_functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 331,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'3-hydroxybutyrate dehydrogenase activity': 1,\n",
       " 'ARF guanyl-nucleotide exchange factor activity': 1,\n",
       " 'ATP binding': 1,\n",
       " \"ATP-dependent 3'-5' DNA helicase activity\": 1,\n",
       " 'DNA binding': 1,\n",
       " 'DNA binding transcription factor activity': 1,\n",
       " 'RNA binding': 1,\n",
       " 'RNA polymerase II activating transcription factor binding': 1,\n",
       " 'RNA polymerase II distal enhancer sequence-specific DNA binding': 1,\n",
       " 'RNA polymerase II proximal promoter sequence-specific DNA binding': 2,\n",
       " 'RNA polymerase II repressing transcription factor binding': 1,\n",
       " 'RNA polymerase II transcription coactivator activity': 1,\n",
       " 'RNA polymerase II transcription cofactor activity': 1,\n",
       " 'RNA polymerase II transcription corepressor activity': 1,\n",
       " 'SNAP receptor activity': 1,\n",
       " 'acetyltransferase activity': 1,\n",
       " 'annealing helicase activity': 1,\n",
       " 'beta-catenin binding': 1,\n",
       " 'bubble DNA binding': 1,\n",
       " 'calcium ion binding': 1,\n",
       " 'chromatin binding': 3,\n",
       " 'extracellular matrix constituent conferring elasticity': 1,\n",
       " 'extracellular matrix structural constituent': 1,\n",
       " 'four-way junction helicase activity': 1,\n",
       " 'helicase activity': 1,\n",
       " 'heparin binding': 1,\n",
       " 'histone acetyltransferase activity': 1,\n",
       " 'hormone activity': 1,\n",
       " 'identical protein binding': 2,\n",
       " 'integrin binding': 1,\n",
       " 'ligand-dependent nuclear receptor transcription coactivator activity': 1,\n",
       " 'metal ion binding': 2,\n",
       " 'oxidized purine DNA binding': 1,\n",
       " 'protein C-terminus binding': 1,\n",
       " 'protein binding': 12,\n",
       " 'protein complex binding': 3,\n",
       " 'protein deacetylase activity': 1,\n",
       " 'protein domain specific binding': 1,\n",
       " 'protein homodimerization activity': 1,\n",
       " 'protein kinase binding': 1,\n",
       " 'receptor activity': 1,\n",
       " 'sequence-specific DNA binding': 2,\n",
       " 'structural molecule activity': 1,\n",
       " 'syntaxin binding': 1,\n",
       " 'telomeric D-loop binding': 1,\n",
       " 'thiol-dependent ubiquitinyl hydrolase activity': 1,\n",
       " 'thyroid hormone receptor binding': 1,\n",
       " 'transcription coactivator activity': 1,\n",
       " 'transcription cofactor activity': 1,\n",
       " 'transcription factor binding': 1,\n",
       " 'transcription regulatory region sequence-specific DNA binding': 1,\n",
       " 'transcriptional activator activity, RNA polymerase II proximal promoter sequence-specific DNA binding': 2,\n",
       " 'ubiquitin protein ligase activity': 1,\n",
       " 'ubiquitin-protein transferase activity': 1,\n",
       " 'vitamin D receptor binding': 1}"
      ]
     },
     "execution_count": 331,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Go process params\n",
    "params = {\n",
    "    'wt': 'json',\n",
    "    'rows': 0,\n",
    "    'q': '*:*',\n",
    "    'facet': 'true',\n",
    "    'facet.sort': 'count',\n",
    "    'facet.method': 'enum',\n",
    "    'facet.mincount': '1',\n",
    "    'facet.limit': '3000',\n",
    "    'facet.field': 'object_label',\n",
    "    'fl': 'subject, subject_label',\n",
    "    'fq': [gene_or_filter,\n",
    "           'relation:\"RO:0002327\"',\n",
    "           \n",
    "          ]\n",
    "}\n",
    "\n",
    "go_process = get_facet_table(solr, params)\n",
    "go_process"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 332,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'AURKA Activation by TPX2': 2,\n",
       " 'Adaptive Immune System': 1,\n",
       " 'Anchoring of the basal body to the plasma membrane': 4,\n",
       " 'Antigen processing: Ubiquitination & Proteasome degradation': 1,\n",
       " 'Autophagy - animal': 1,\n",
       " 'Butanoate metabolism': 1,\n",
       " 'Cell Cycle': 2,\n",
       " 'Cell Cycle, Mitotic': 2,\n",
       " 'Centrosome maturation': 2,\n",
       " 'Chromatin modifying enzymes': 1,\n",
       " 'Chromatin organization': 1,\n",
       " 'Cilium Assembly': 4,\n",
       " 'Class I MHC mediated antigen processing & presentation': 1,\n",
       " 'Degradation of the extracellular matrix': 1,\n",
       " 'Developmental Biology': 1,\n",
       " 'Elastic fibre formation': 1,\n",
       " 'Epigenetic regulation of gene expression': 1,\n",
       " 'Extracellular matrix organization': 1,\n",
       " 'Factors involved in megakaryocyte development and platelet production': 1,\n",
       " 'Fatty acid, triacylglycerol, and ketone body metabolism': 3,\n",
       " 'G2/M Transition': 2,\n",
       " 'Gene Expression': 2,\n",
       " 'Generic Transcription Pathway': 1,\n",
       " 'HATs acetylate histones': 1,\n",
       " \"Hedgehog 'off' state\": 1,\n",
       " 'Hemostasis': 1,\n",
       " \"Huntington's disease\": 1,\n",
       " 'Immune System': 3,\n",
       " 'Innate Immune System': 2,\n",
       " 'Integrin cell surface interactions': 1,\n",
       " 'Intra-Golgi and retrograde Golgi-to-ER traffic': 1,\n",
       " 'Intra-Golgi traffic': 1,\n",
       " 'Ketone body metabolism': 1,\n",
       " 'Loss of Nlp from mitotic centrosomes': 2,\n",
       " 'Loss of proteins required for interphase microtubule organization from the centrosome': 2,\n",
       " 'Membrane Trafficking': 1,\n",
       " 'Metabolic pathways': 1,\n",
       " 'Metabolism': 3,\n",
       " 'Metabolism of lipids and lipoproteins': 3,\n",
       " 'Mitotic G2-G2/M phases': 2,\n",
       " 'Molecules associated with elastic fibres': 1,\n",
       " 'Negative epigenetic regulation of rRNA expression': 1,\n",
       " 'Neutrophil degranulation': 2,\n",
       " 'NoRC negatively regulates rRNA expression': 1,\n",
       " 'Organelle biogenesis and maintenance': 4,\n",
       " 'PPARA activates gene expression': 1,\n",
       " 'Post-translational protein phosphorylation': 1,\n",
       " 'RUNX1 regulates genes involved in megakaryocyte differentiation and platelet function': 1,\n",
       " 'RUNX2 regulates osteoblast differentiation': 1,\n",
       " 'Recruitment of NuMA to mitotic centrosomes': 2,\n",
       " 'Recruitment of mitotic centrosome proteins and complexes': 2,\n",
       " 'Regulation of Insulin-like Growth Factor (IGF) transport and uptake by Insulin-like Growth Factor Binding Proteins (IGFBPs)': 1,\n",
       " 'Regulation of PLK1 Activity at G2/M Transition': 2,\n",
       " 'Regulation of lipid metabolism by Peroxisome proliferator-activated receptor alpha (PPARalpha)': 2,\n",
       " 'SNARE interactions in vesicular transport': 1,\n",
       " 'SUMOylation of chromatin organization proteins': 1,\n",
       " 'Signal Transduction': 1,\n",
       " 'Signaling by Hedgehog': 1,\n",
       " 'Synthesis and degradation of ketone bodies': 1,\n",
       " 'Synthesis of Ketone Bodies': 1,\n",
       " 'Thyroid hormone signaling pathway': 2,\n",
       " 'Transcriptional misregulation in cancer': 1,\n",
       " 'Transcriptional regulation of white adipocyte differentiation': 1,\n",
       " 'Ubiquitin mediated proteolysis': 1,\n",
       " 'Utilization of Ketone Bodies': 1,\n",
       " 'Vesicle-mediated transport': 1}"
      ]
     },
     "execution_count": 332,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# pathway params\n",
    "params = {\n",
    "    'wt': 'json',\n",
    "    'rows': 0,\n",
    "    'q': '*:*',\n",
    "    'facet': 'true',\n",
    "    'facet.sort': 'count',\n",
    "    'facet.method': 'enum',\n",
    "    'facet.mincount': '1',\n",
    "    'facet.limit': '3000',\n",
    "    'facet.field': 'object_label',\n",
    "    'fl': 'subject, subject_label',\n",
    "    'fq': [gene_or_filter,\n",
    "           'object_category:\"pathway\"',\n",
    "           \n",
    "          ]\n",
    "}\n",
    "\n",
    "pathways = get_facet_table(solr, params)\n",
    "pathways"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# TODO GO term enrichment with ontobio"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
